### PR TITLE
As a user i can perform docker search of docker v2 content

### PR DESCRIPTION
**Pull request by aprajshekhar**
Added :

- Filter for V2 metadata
- Logic to run both filters and aggregate the results removing the duplicates

### CODE REVIEW

**Comment by ipanova** [comment]
@aprajshekhar i took a quick glance at this and i did not notice that you added the endpoint for v2 content?
i don't think it sounds correct to access `v1/search` and still being able to query v2 content.
can you also elaborate on why you decided not to add gsa backend? you mentioned on the email that gsa does not provide filtering? at least we say in docs that both backends are configurable so i guess we should add gsa support for v2 as well.
i am new to solr and gsa so i am getting familiar to those just now.
once you finish the code and tests, some docs would need to be added, same for release notes.

Thank you for your PR!

**Comment by ipanova** [comment]
@aprajshekhar a quick update, looks like noone uses gsa, so it is ok to have just solr backend for v2, i am going to update the deliverables for this story

**Comment by aprajshekhar** [comment]
@ipanova, since registry API supports search via only v1/search, adding a v2/search would not work as docker client would not call it. I had tried adding v2/search and it never got hit. On registry.access.redhat.com, v1/search is providing v2 content results. Hope this helps.

File: crane/search/base.py [hunk 1]
```diff
@@ -92,7 +110,7 @@ def _get_data(url):
         """
         try:
             # one second timeout
-            response = urllib2.urlopen(url, timeout=1)
+            response = urllib2.urlopen(url, timeout=30)
```
Comment (ipanova) [inline crane/search/base.py]: any reason you bumped to such a bit timeout in comparison the the old one?

Comment (aprajshekhar) [reply crane/search/base.py]: The Solr takes more than 1 second in some cases. resulting in 503. Hence. I had bumped it up.

File: crane/search/base.py [hunk 2]
```diff
@@ -64,6 +64,24 @@ def _filter_result(self, result):
...
+        both known by this app (aka we have it in the app data), and if the user
+        is authorized to access it.
+
+        :param result:  one search result
+        :type  result:  SearchResult
+        :return:    True iff the repository is known and the user is authorized,
+                    else False
+        :rtype:     bool
+        """
+        try:
```
Comment (ipanova) [inline crane/search/base.py]: what about keeping the just the _filter_result() and instead:

``` 
        try:
            app_util.name_is_authorized(result.name)
        except exceptions.HTTPError:
            try:
                app_util.repo_is_authorized(result.name)
            except exceptions.HTTPError:
                return False
        return True
```

or

```
       try:
            if  not app_util.name_is_authorized(result.name):
                return True
           app_util.repo_is_authorized(result.name)
        except exceptions.HTTPError:
            return False
        return True

This solution will simplify: https://github.com/pulp/crane/pull/81/files#diff-aaa7c890485705f9a4ef1adebdd8784dR47
You would not need to call the filter_result twice on the results.
The protection comes per repo and not per type/version of the content.

And you also would be able to get rid of https://github.com/pulp/crane/pull/81/files#diff-aaa7c890485705f9a4ef1adebdd8784dR105

Comment (mhrivnak) [reply crane/search/base.py]: I agree that these should be a single function that does as many checks as necessary to get its job done. A great indicator is that the doc blocks are identical.

**Comment by aprajshekhar** [comment]
@ipanova, @mhrivnak I have made changes as suggested.

Just so that we are on the same page when V2 was initially implemented (2015-2016) (for which I had created a PR), the pattern suggested to me was to have a new method for all V2 related logic. Hence, when I implemented search for registry.access.redhat.com, I followed the same pattern.
Also, I know that protection comes per repo :). As I had to make it work when registry.access.redhat.com needed to  work with entitlements.

File: crane/search/base.py [hunk 3]
```diff
@@ -59,9 +59,14 @@ def _filter_result(self, result):
         :rtype:     bool
         """
         try:
+            # check against repositories in V1 data dict
```
Comment (ipanova) [inline crane/search/base.py]: i suggest to swap the v1 check with the v2 check first, because most of the repos are v2 nowadays

**Comment by ipanova** [comment]
@aprajshekhar  I think we are on the same page :) Sorry for the confusion, maybe that pattern which was suggested 2 years ago was reasonable with uncertainty at that moment about docker distribution bringing over search in v2 api or not.
Thank you for the updates.

File: docs/index.rst [hunk 4]
```diff
@@ -105,7 +106,8 @@ The API supporting ``docker search`` can be enabled by configuring a Solr
 deployment for use by crane. In crane's configuration file, a section
 ``[solr]`` must exist with key ``url``. The URL will be used in a GET request,
 and it must contain the string ``{0}`` as a placeholder where the search string
-will be inserted.
+will be inserted. The search returns both results containing both V1
```
Comment (ipanova) [inline docs/index.rst]: can you please add a note that search for v2 works just with Solr.

Comment (aprajshekhar) [reply docs/index.rst]: Solr works with V1 and V2. GSA works with V1 only.

File: docs/index.rst [hunk 5]
```diff
@@ -273,7 +275,8 @@ The 3.0.0 release of Crane adds support for the Docker v2 API manifest lists sch
 Crane now supports version 4 of the json data file produced by Pulp. The new version enables
 Crane to serve manifest lists. In order to decide to which manifest type to redirect
 the request, Crane looks for an optional accept header specified in the request and based on this
-information redirects whether to the manifest list or image manifest.
+information redirects whether to the manifest list or image manifest. The search will also return results
```
Comment (ipanova) [inline docs/index.rst]: This should under 3.1.0 version

Please create a new paragraph where you'd mention the v2 docker search.

File: docs/index.rst [hunk 6]
```diff
@@ -105,7 +106,8 @@ The API supporting ``docker search`` can be enabled by configuring a Solr
 deployment for use by crane. In crane's configuration file, a section
 ``[solr]`` must exist with key ``url``. The URL will be used in a GET request,
 and it must contain the string ``{0}`` as a placeholder where the search string
-will be inserted.
+will be inserted. The search returns both results containing both V1
+as well as V2 versions.
```
Comment (ipanova) [inline docs/index.rst]: s/versions/docker content

File: docs/index.rst [hunk 7]
```diff
@@ -83,7 +83,8 @@ GSA
 The API supporting ``docker search`` can be enabled by configuring a Google
 Search Appliance for use by crane. In crane's configuration file, a section
 ``[gsa]`` must exist with key ``url``. The URL will be used in a GET request,
-and a query parameter ``q`` will be added with a search term.
+and a query parameter ``q`` will be added with a search term. This is available
+only for V1 content type.
```
Comment (ipanova) [inline docs/index.rst]: s/content type/docker content

**Comment by ipanova** [comment]
@aprajshekhar back to you

**Comment by aprajshekhar** [comment]
I have implemented the suggested changes.

**Pull request merged**